step/step.go

```diff
@@ -0,0 +1,255 @@
+
+	if len(paths) == 0 {
+		r.logger.Printf("No files found.")
+
+		return Result{
+			HtmlReportDir: "",
+		}, nil
+	}
+
+	r.logger.Printf("List of files:")
+	for _, path := range paths {
+		r.logger.Printf("- %s", path)
+	}
+
+	rootDir, err := r.htmlReportsRootDir()
+	if err != nil {
+		return Result{}, fmt.Errorf("failed to create test report directory: %w", err)
+	}
+
+	r.logger.Println()
+	r.logger.Infof("Generating reports")
+
+	for _, path := range paths {
+		if err := r.generateTestReport(rootDir, path); err != nil {
+			r.logger.Errorf("failed to generate test report (%s): %w", path, err)
+		}
+	}
+
+	r.logger.Println()
+	r.logger.Donef("Finished")
+
+	return Result{
+		HtmlReportDir: rootDir,
+	}, nil
+}
+
+func (r *ReportGenerator) Export(result Result) error {
+	return r.exporter.ExportOutput(htmlReportDirKey, result.HtmlReportDir)
+}
+
+func (r *ReportGenerator) generateTestReport(rootDir string, xcresultPath string) error {
+	baseName := strings.TrimSuffix(filepath.Base(xcresultPath), filepath.Ext(xcresultPath))
+	dirPath := filepath.Join(rootDir, baseName)
+	err := os.Mkdir(dirPath, 0755)
+	if err != nil {
+		if os.IsExist(err) {
+			r.logger.Warnf("Html report already exists for %s at %s", baseName, dirPath)
+			return nil
+		}
+		return err
+	}
+
+	r.logger.Printf("Generating report for: %s", baseName)
+	params := []string{"run", "bitrise-io/XCTestHTMLReport@speed-improvements", "--output", dirPath, xcresultPath}
+	cmd := r.commandFactory.Create("mint", params, nil)
+	_, err = cmd.RunAndReturnTrimmedCombinedOutput()
+	if err != nil {
+		return fmt.Errorf("failed to export html: %w", err)
+	}
+
+	if err := moveAssets(xcresultPath, dirPath); err != nil {
+		return fmt.Errorf("failed to move assets: %w", err)
+	}
+
+	return nil
+}
+
+func (r *ReportGenerator) htmlReportsRootDir() (string, error) {
+	reportDir := r.envRepository.Get(htmlReportDirKey)
+	if reportDir == "" {
+		return os.MkdirTemp("", "html-reports")
+	}
+
+	exists, err := pathutil.NewPathChecker().IsDirExists(reportDir)
+	if err != nil {
+		return "", err
+	}
+	if !exists {
+		return "", fmt.Errorf("html report dir (%s) does not exist or is not a folder", reportDir)
+	}
+
+	return reportDir, nil
+}
+
+func collectFilesWithPatterns(patterns []string) ([]string, error) {
+	// Go does not have a set, so a map will help filter out duplicate results.
+	allMatches := map[string]struct{}{}
+
+	for _, pattern := range patterns {
+		matches, err := zglob.Glob(pattern)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, match := range matches {
+			allMatches[match] = struct{}{}
+		}
+	}
+
+	var paths []string
+	for key := range allMatches {
+		paths = append(paths, key)
+	}
+
+	return paths, nil
+}
+
+func moveAssets(xcresultPath string, htmlReportDir string) error {
+	entries, err := os.ReadDir(xcresultPath)
+	if err != nil {
+		return err
+	}
+
+	assetFolder := filepath.Join(htmlReportDir, filepath.Base(xcresultPath))
+	if err := os.Mkdir(assetFolder, 0755); err != nil {
```

Comment:
Note: permission 0755 seems generous, could it be 0750 or 0700?

Reply:
That is idea. I will update this in the next PR.